[RDY] Remove term modules.

tarruda opened the pull request:
Continuing the work of #1820

justinmk commented:
Will the removal of os_unix wait until after #1820?

tarruda commented:
> Will the removal of os_unix wait until after #1820?

Yes, this PR is a continuation of #1820

Pyrohh commented:
Do you know if `ttymouse` will still be used? My hope is that the mouse handling scheme will be automatically detected.

tarruda commented:
> Do you know if ttymouse will still be used? My hope is that the mouse handling scheme will be automatically detected.

Yes. I plan to enhance terminal detection features in tui.c removing the need for these options.

tarruda commented:
> Will the removal of os_unix wait until after #1820?

I only removed some functions from os_unix, will leave it alone for now

tarruda committed:
refactor: Remove term modules and termcap options

- Removed term.c, term.h and term_defs.h
- Tests for T_* values were removed. screen.c was simplified as a
  consequence(the best strategy for drawing is implemented in the UI layer)
- Redraw functions now call ui.c functions directly. Updates are flushed with
  `ui_flush()`
- Removed all termcap options(they now return empty strings for compatibility)
- &term/&ttybuiltin options return a constant value(nvim)
- &t_Co is still available, but it mirrors t_colors directly
- Remove cursor tracking from screen.c and the `screen_start` function. Now the
  UI is expected to maintain cursor state across any call, and reset it when
  resized.
- Remove unused code

Happy-Dude commented:
@tarruda, today I noticed that my conditional for checking what kind of terminal (n)vim is started in,

```viml
if ... || ((&term =~ "xterm") (&term =~ "screen") || (&term =~ "color"))
```

is not longer working. I was wondering if this change was known and if there were any plans on reimplementing this kinda of detection?

justinmk commented:
@Happy-Dude what about `$TERM`? And is hand-rolled terminal detection a good thing? :)

Happy-Dude commented:
@justinmk -- I guess I stuck with that method because it has always worked. Is it better to use $TERM?

tarruda commented:
@Happy-Dude `&term` no longer makes sense in nvim, considering its client-server architecture(there can be multiple UIs connected at the same time, some of them which may not even be terminals). The option is left to avoid vimscript errors but it returns a fixed value of `"nvim"`. For most scenarios, using `$TERM` will work though.

fwalch commented:
@tarruda Would it make sense to implement a way for GUIs to hook into the retrieval of &term and return an appropriate value, e.g. `$TERM` for the TUI, `pynvim` for the Python GUI, ...?

Happy-Dude commented:
@fwalch -- I like that! That's a very clever workaround.
Though that would be a very different from the original intention of term (from `:help term`):

	Name of the terminal.  Used for choosing the terminal control
	characters.  Environment variables are expanded |:set_env|.

Maybe it's worth a larger discussion?

tarruda commented:
> @tarruda Would it make sense to implement a way for GUIs to hook into the retrieval of &term and return an appropriate value, e.g. $TERM for the TUI, pynvim for the Python GUI, ...? Or just return $TERM by default instead of nvim (and maybe the value that gvim uses if $TERM is empty)?

The value of `&term` was previously populated from `$TERM` value, and for the most common case which is one UI running in the same process, inspecting `$TERM` instead of  `&term` will work as expected.

As for hooking into UI to provide the value, its not possible because there can be multiple UIs attached, some which are not even terminals. What value to return in this case?

Another problem is `&t_Co`. Right now it is set to a value of 256 colors since thats the most common case, and it can be overriden by the user. But it won't account for the case where two terminals are attached, and one of them don't support 256 colors.

One solution that would work for most cases is to let new tui module set `&term` and `&t_Co` on startup.

Happy-Dude commented:
@tarruda does &term and &t_Co have to be the same for an "instance" of nvim even when multiple UIs are attached? Would it be more beneficial for term and t_Co to be context/ UI based instead? 

If I connect to nvim via xterm and atom, can term just respond `xterm` and `atom` (respectively) when called during that connected UI? (Sorry, unsure of the terminology for nvim's new UI architecture.) I guess what I'm trying so ask is whether or not term and t_Co need to be static during a running instance of nvim.

Oberon00 commented:
It seems like UIs might need a new variable space (like `b:` for buffers or `w:` for windows) along with a notion of which UI is currently active (last used?).

tarruda commented:
> @tarruda does &term and &t_Co have to be the same for an "instance" of nvim even when multiple UIs are attached? Would it be more beneficial for term and t_Co to be context/ UI based instead?

I can't make vimscript options UI-specific, but I'd like to solve that by allowing arbitrary attributes to be passed to the highlight command, `:hi term88fg=5 term256fg=16  guifg=#f3f3f3` for example. This way it would be possible to develop color schemes that work on a variety of UIs.

> If I connect to nvim via xterm and atom, can term just respond xterm and atom (respectively) when called during that connected UI? (Sorry, unsure of the terminology for nvim's new UI architecture.) I guess what I'm trying so ask is whether or not term and t_Co need to be static during a running instance of nvim.

I don't see how that would be possible without extremely hacky and error-prone code. All UIs send their input to the same buffer, and its hard to determine when the input will actually be evaluated.

> It seems like UIs might need a new variable space (like b: for buffers or w: for windows) along with a notion of which UI is currently active (last used?).

The API function used to queue input will return immediately(before the keys are even processed)

tarruda commented:
@elmart /@oni-link / @Tranquility could you give this branch a spin? It cleans a lot of stuff left by #1820.

elmart commented:
Sure. You mean working with it to see if there are undesired effects, don't you?
I'll be doing it this evening.

tarruda committed:
tui: Set setaf/setab capabilities for TERM=xterm

Many common terminals that set TERM=xterm and $COLORTERM support 256 colors. If
this is detected, use force the hardcoded xterm's setaf/setab capabilities.

Tranquility commented:
I tested it and found one problem. The version of this branch segfaults if I try to use fugitives patch feature (:Gstatus and press <kbd>p</kbd> on top of a changed file). I know it doesn't work in nvim anyway, but despite that it shouldn't segfault right?

tarruda commented:
> I tested it and found one problem. The version of this branch segfaults if I try to use fugitives patch feature (:Gstatus and press p on top of a changed file). I know it doesn't work in nvim anyway, but despite that it shouldn't segfault right?

:+1: will investigate, thanks

tarruda committed:
tui: Remove option cache

tarruda committed:
tui: Use unibi_format instead of unibi_run

The current will segfault for large chunks of output because the output buffer
will be overrun.

Using unibi_format is simple because we can simply flush the buffer when its
full.

tarruda commented:
@Tranquility the last commit should fix the problem

elmart commented:
Is this ready to test? It seems like there are too many failed builds.

tarruda commented:
> Is this ready to test? It seems like there are too many failed builds.

Its ready, these failures were caused by a small bug introduced in the last commit. Should be fixed now

Pyrohh commented on src/nvim/main.c:
```diff
@@ -2115,3 +2093,27 @@ static void check_swap_exists_action(void)
+  ui_flush();
+
+#ifdef MACOS_CONVERT
+  mac_conv_init();
```
This was removed with #1979.

tarruda replied:
:+1:

Pyrohh commented on src/nvim/memory.c:
```diff
@@ -149,7 +149,7 @@ void *xmallocz(size_t size)
   size_t total_size = size + 1;
   if (total_size < size) {
-    OUT_STR(_("Vim: Data too large to fit into virtual memory space\n"));
+    mch_errmsg(_("Vim: Data too large to fit into virtual memory space\n"));
```
Should probably be updated to nvim.

tarruda replied:
Since most of these messages are still prefixed with "Vim:", I'd rather leave it like this for now and change in another PR

Pyrohh replied:
:+1:

On February 21, 2015 2:58:56 AM EST, Thiago de Arruda <notifications@github.com> wrote:
>> @@ -149,7 +149,7 @@ void *xmallocz(size_t size)
>>  {
>>    size_t total_size = size + 1;
>>    if (total_size < size) {
>> -    OUT_STR(_("Vim: Data too large to fit into virtual memory
>space\n"));
>> +    mch_errmsg(_("Vim: Data too large to fit into virtual memory
>space\n"));
>
>Since most of these messages are still prefixed with "Vim:", I'd rather
>leave it like this for now and change in another PR
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/neovim/neovim/pull/1997/files#r25120696

Pyrohh commented on src/nvim/misc1.c:
```diff
@@ -3313,17 +3311,16 @@ void preserve_exit(void)
-      OUT_STR("Vim: preserving files...\n");
-      screen_start();               // don't know where cursor is now
-      out_flush();
+      mch_errmsg((uint8_t *)"Vim: preserving files...\n");
```
same here

Pyrohh commented on src/nvim/misc1.c:
```diff
@@ -3331,7 +3328,7 @@ void preserve_exit(void)
   ml_close_all(false);              // close all memfiles, without deleting
 
-  OUT_STR("Vim: Finished.\n");
+  mch_errmsg("Vim: Finished.\n");
```
and here.

Pyrohh commented on src/nvim/ui.c:
```diff
@@ -527,3 +463,27 @@ static void flush_cursor_update(void)
+{
+  static int showing_insert_mode = MAYBE;
+
+  if (!full_screen)
```
brackets are missing

tarruda replied:
:+1:

Pyrohh commented on src/nvim/main.c:
```diff
@@ -2115,3 +2093,27 @@ static void check_swap_exists_action(void)
+static void term_init(void)
+{
+  Columns = 80;
+  Rows = 24;
```
Why set these two variables? I ask because nvim on this branch seems to work fine in a tmux pane smaller than these dimensions, and when I remove these two lines all tests still pass and everything still seems to work fine.

tarruda replied:
These are already the values initialized in globals.h so its not necessary :+1: 

I simply removed term_init and replaced by a call to `setbuf`(ui_flush is not required at this point)
